// angular-notifier/esm2015/lib/animation-presets/slide.animation-preset.js
const ɵ0 = (notification) => {
    // Prepare variables
    const config = notification.component.getConfig();
    const shift = notification.component.getShift();
    let from;
    let to;
    // Configure variables, depending on configuration and component
    if (config.position.horizontal.position === 'left') {
        from = {
            transform: `translate3d( 0, ${shift}px, 0 )`,
        };
        to = {
            transform: `translate3d( calc( -100% - ${config.position.horizontal.distance}px - 10px ), ${shift}px, 0 )`,
        };
    }
    else if (config.position.horizontal.position === 'right') {
        from = {
            transform: `translate3d( 0, ${shift}px, 0 )`,
        };
        to = {
            transform: `translate3d( calc( 100% + ${config.position.horizontal.distance}px + 10px ), ${shift}px, 0 )`,
        };
    }
    else {
        let horizontalPosition;
        if (config.position.vertical.position === 'top') {
            horizontalPosition = `calc( -100% - ${config.position.horizontal.distance}px - 10px )`;
        }
        else {
            horizontalPosition = `calc( 100% + ${config.position.horizontal.distance}px + 10px )`;
        }
        from = {
            transform: `translate3d( -50%, ${shift}px, 0 )`,
        };
        to = {
            transform: `translate3d( -50%, ${horizontalPosition}, 0 )`,
        };
    }
    // Done
    return {
        from,
        to,
    };
}, ɵ1 = (notification) => {
    // Prepare variables
    const config = notification.component.getConfig();
    let from;
    let to;
    // Configure variables, depending on configuration and component
    if (config.position.horizontal.position === 'left') {
        from = {
            transform: `translate3d( calc( -100% - ${config.position.horizontal.distance}px - 10px ), 0, 0 )`,
        };
        to = {
            transform: 'translate3d( 0, 0, 0 )',
        };
    }
    else if (config.position.horizontal.position === 'right') {
        from = {
            transform: `translate3d( calc( 100% + ${config.position.horizontal.distance}px + 10px ), 0, 0 )`,
        };
        to = {
            transform: 'translate3d( 0, 0, 0 )',
        };
    }
    else {
        let horizontalPosition;
        if (config.position.vertical.position === 'top') {
            horizontalPosition = `calc( -100% - ${config.position.horizontal.distance}px - 10px )`;
        }
        else {
            horizontalPosition = `calc( 100% + ${config.position.horizontal.distance}px + 10px )`;
        }
        from = {
            transform: `translate3d( -50%, ${horizontalPosition}, 0 )`,
        };
        to = {
            transform: 'translate3d( -50%, 0, 0 )',
        };
    }
    // Done
    return {
        from,
        to,
    };
};
/**
 * Slide animation preset
 */
export const slide = {
    hide: ɵ0,
    show: ɵ1,
};
export { ɵ0, ɵ1 };
